src/client/Microsoft.Identity.Client/AuthScheme/PoP/PopBindingResolver.cs

```diff
@@ -0,0 +1,113 @@
+            {
+                if (certCred.Certificate == null)
+                {
+                    throw new MsalClientException(
+                        MsalError.MtlsCertificateNotProvided,
+                        MsalErrorMessage.MtlsCertificateNotProvidedMessage);
+                }
+
+                return;
+            }
+
+            // ────────────────────────────────────
+            // Case 2 – Client‑assertion delegate
+            // ────────────────────────────────────
+            if (serviceBundle.Config.ClientCredential is ClientAssertionDelegateCredential cadc)
+            {
+                var opts = new AssertionRequestOptions
+                {
+                    ClientID = serviceBundle.Config.ClientId,
+                    ClientCapabilities = serviceBundle.Config.ClientCapabilities,
+                    Claims = commonParameters.Claims,
+                    CancellationToken = ct
+                };
+
+                ClientAssertion ar = await cadc.GetAssertionAsync(opts, ct).ConfigureAwait(false);
+
+                if (ar.TokenBindingCertificate == null)
+                {
+                    throw new MsalClientException(
+                        MsalError.MtlsCertificateNotProvided,
+                        MsalErrorMessage.MtlsCertificateNotProvidedMessage);
+                }
+
+                Wire(commonParameters, ar.TokenBindingCertificate, serviceBundle);
+                return;
+            }
+
+            // ────────────────────────────────────
+            // Case 3 – Any other credential (client‑secret etc.)
+            // ────────────────────────────────────
+            throw new MsalClientException(
+                MsalError.MtlsCertificateNotProvided,
+                MsalErrorMessage.MtlsCertificateNotProvidedMessage);
+        }
+
+        /// <summary>
+        /// Common wiring + region check.
+        /// </summary>
+        private static void Wire(
+            AcquireTokenCommonParameters commonParameters,
+            X509Certificate2 cert,
+            IServiceBundle serviceBundle)
+        {
+            // Region requirement (AAD only)
+            if (serviceBundle.Config.Authority.AuthorityInfo.AuthorityType == AuthorityType.Aad &&
+                serviceBundle.Config.AzureRegion == null)
+            {
+                throw new MsalClientException(
+                    MsalError.MtlsPopWithoutRegion,
+                    MsalErrorMessage.MtlsPopWithoutRegion);
+            }
+
+            commonParameters.AuthenticationOperation = new MtlsPopAuthenticationOperation(cert);
+            commonParameters.MtlsCertificate = cert;
+
+            commonParameters.CacheKeyComponents ??= new SortedList<string, string>(StringComparer.Ordinal);
+
+            commonParameters.CacheKeyComponents[Constants.CertSerialNumber] = cert.SerialNumber;
+
+            serviceBundle.Config.CertificateIdToAssociateWithToken = cert.Thumbprint;
```

Comment:
I recommend we use SHA-2 thumbprint. SHA1 thumbprint is banned. Just have an extension method on X509Certificate2 like GetSha2Thumbprint()